atc/exec/garden_factory.go

```diff
@@ -150,13 +150,13 @@ func (factory *gardenFactory) Task(
 		// external task - construct a source which reads it from file
 		taskConfigSource = FileConfigSource{ConfigPath: plan.Task.ConfigPath}
 
-		// use 'vars' from the pipeline + cred mgr variables for interpolation
+		// for interpolation - use 'vars' from the pipeline, and then fill remaining with cred mgr variables
 		taskVars = []boshtemplate.Variables{boshtemplate.StaticVariables(plan.Task.Vars), credMgrVariables}
```

Comment:
Hmm are you sure this shouldn't just be reversed? For example, the current semantics with `-l` are that the right-most side wins, even with `((vars))` in a pipeline:

```sh
~/w/concourse (fix-smoke) $ cat booklit.yml 
resources:
- name: booklit
  type: git
  source:
    uri: https://github.com/vito/booklit
    branch: ((branch))

jobs:
- name: unit
  plan:
  - get: booklit
  - task: test
    file: booklit/ci/test.yml

~/w/concourse (fix-smoke) $ cat develop.yml 
branch: develop

~/w/concourse (fix-smoke) $ cat master.yml 
branch: master

~/w/concourse (fix-smoke) $ fly -t l validate-pipeline -o -c booklit.yml -l develop.yml -l master.yml
jobs:
- name: unit
  plan:
  - get: booklit
  - file: booklit/ci/test.yml
    task: test
resources:
- name: booklit
  source:
    branch: master
    uri: https://github.com/vito/booklit
  type: git
```

The mental model is that they are appended and stacked on top of each other. So technically what's listed first is lowest precedence.

Reply:
Yes. The command line semantics doesn't change - whatever specified on the right side wins. We want to preserve that. 

But, in the source code, I always felt a normal order should win:
* when you iterate over a slice or array, the default is to iterate in natural order and break from the loop if a match is found
* this is how boshtemplate.NewMultiVars(), for example, is implemented. so I always felt that whatever we pass into NewTemplateResolver() should not be reversed before passing to boshtemplate.NewMultiVars()
* and when I wrote `[]boshtemplate.Variables{boshtemplate.StaticVariables(plan.Task.Vars), credMgrVariables}`, the original intention was that plans vars should be tried first.

So I am trying to stick to my original plan. But if you strongly feel that we should follow "stacked on top of each other" approach in the source code too, I can make that change and change all comments to make sure the interface is clear.

Reply:
P.S. I don't think this is going to solve the Vault slowness issue though. This only affects tasks with FileConfigSource, when an array with more than one element is passed to NewTemplateResolver(). Vault issue happens on regular tasks too, where an array consists of just one element with credential manager vars.

Reply:
Sounds fine, just wanted to make sure this didn't bubble up and change the existing CLI behavior. :+1: